### PR TITLE
Restrictive Federation

### PR DESCRIPTION
**Pull request by harveyxia**
BOT NOTES: 
resolves https://github.com/solo-io/gloo-mesh/issues/804

### CODE REVIEW

File: pkg/mesh-networking/apply/applier.go [hunk 1]
```diff
@@ -680,6 +680,62 @@ func getAppliedVirtualMesh(
...
+				}
+			}
+			// federate to specified Meshes
+			for _, meshRef := range federationSelector.GetMeshes() {
+				federatedToMeshes.Insert(meshRef)
+			}
+		}
+	default:
+		// no federation applied
+		return nil
```
Comment (EItanya) [inline pkg/mesh-networking/apply/applier.go:721]: Since the default behavior used to be `permissive`, do we want to maintain parity with that?

Comment (harveyxia) [reply pkg/mesh-networking/apply/applier.go:721]: will raise this in parking lot tomorrow, I don't have strong opinions either way

Comment (harveyxia) [reply pkg/mesh-networking/apply/applier.go]: In today's discussion we decided to go with this approach:

1. Deprecate the existing `mode` field.
2. Honor the deprecated `mode` field only if the new `selectors` field is not populated.
3. Permissive federation semantics are expressed with the following:

```yaml
spec:
    federation:
      selectors:
      - {}
```

File: api/networking/v1/virtual_mesh.proto [hunk 2]
```diff
@@ -105,14 +105,16 @@ message VirtualMeshSpec {
     // originating from Workloads external to the Destination's Mesh.
     message Federation {
 
+        // Selectively federate Destinations to specific external meshes.
+        // If omitted, all Destinations will be federated to all Meshes in the VirtualMesh.
```
Comment (EItanya) [inline api/networking/v1/virtual_mesh.proto:109]: I think we should be specific about the different states, eg.
1. nil = no federation
2. empty = everything
3. selectors

File: pkg/mesh-networking/apply/applier.go [hunk 3]
```diff
@@ -688,37 +688,33 @@ func getFederatedToMeshes(
...
-	case *networkingv1.VirtualMeshSpec_Federation_Permissive:
-		// permissive federation exposes the Destination to all Meshes in the VirtualMesh
-		for _, groupedMeshRef := range virtualMesh.Spec.GetMeshes() {
-			federatedToMeshes.Insert(groupedMeshRef)
-		}
-	case *networkingv1.VirtualMeshSpec_Federation_Restrictive:
-		// omitted federation selectors have permissive semantics
-		if len(mode.Restrictive.GetFederationSelectors()) < 1 {
+	// respect deprecated `mode` field only if new federation selectors are not specified
+	if len(virtualMesh.Spec.GetFederation().GetSelectors()) < 1 {
```
Comment (EItanya) [inline pkg/mesh-networking/apply/applier.go:692]: super nit: 
```suggestion
	if len(virtualMesh.Spec.GetFederation().GetSelectors()) == 0 {
```

File: api/networking/v1/virtual_mesh.proto [hunk 4]
```diff
@@ -105,14 +105,16 @@ message VirtualMeshSpec {
...
+        repeated FederationSelector selectors = 4;
+
+        // DEPRECATED: Use `selectors` instead. Omission of the `selectors` field has permissive semantics.
```
Comment (EItanya) [inline api/networking/v1/virtual_mesh.proto:112]: Will this show up in the docs?
If not we should deprecate the `permissive` field below

File: api/networking/v1/virtual_mesh.proto (hunk 2 above)
Comment (ilackarms) [reply api/networking/v1/virtual_mesh.proto:109]:
```yaml
# disable federation
selectors: null
```

```yaml
# restricted federation
selectors: 
- destinationSelectors:
  - kubeServiceMatcher:
      labels:  
        app: reviews
      clusters:
      - remote-cluster
  meshes:
  - name: istiod-istio-system-mgmt-cluster 
    namespace: gloo-mesh
```

```yaml
# permissive federation
selectors: 
- destinationSelectors: null
  meshes: null

# can also be expressed as
selectors: 
- {}
```

Comment (EItanya) [reply api/networking/v1/virtual_mesh.proto:109]: Spoke with @harveyxia offline and this sounds ok

File: pkg/mesh-networking/apply/applier.go [hunk 5]
```diff
@@ -317,6 +317,9 @@ func validateAndReturnApprovedFederation(
 	reporter *applyReporter,
 	destination *discoveryv1.Destination,
 ) *discoveryv1.DestinationStatus_AppliedFederation {
+	if destination.Status.AppliedFederation == nil {
```
Comment (ilackarms) [inline pkg/mesh-networking/apply/applier.go:320]: jw, why the change here?

File: docs/content/guides/federate_identity.md [hunk 6]
```diff
@@ -153,7 +153,10 @@ spec:
     shared:
       rootCertificateAuthority:
         generated: {}
-  federation: {}
+  federation:
+    # federate all Destinations to all external meshes
+    selectors:
+    - {}
```
Comment (ilackarms) [inline docs/content/guides/federate_identity.md:159]: just wondering, why the change here?

File: pkg/mesh-networking/apply/applier.go (hunk 5 above)
Comment (harveyxia) [reply pkg/mesh-networking/apply/applier.go:320]: without this clause, a Destination's `status.AppliedFederation` will be populated even though it's not actually being federated to any external meshes

File: docs/content/guides/federate_identity.md (hunk 6 above)
Comment (harveyxia) [reply docs/content/guides/federate_identity.md:159]: updating our docs to use the non-deprecated field

File: pkg/mesh-networking/apply/applier.go [hunk 7]
```diff
@@ -680,6 +680,58 @@ func getAppliedVirtualMesh(
...
+func getFederatedToMeshes(
+	destination *discoveryv1.Destination,
+	parentMesh *v1.ObjectRef,
+	virtualMesh *networkingv1.VirtualMesh,
+) []*v1.ObjectRef {
+	federatedToMeshes := sets.NewResourceSet()
+
+	// respect deprecated `mode` field only if new federation selectors are not specified
+	if len(virtualMesh.Spec.GetFederation().GetSelectors()) < 1 {
+		switch virtualMesh.Spec.GetFederation().GetMode().(type) {
```
Comment (ilackarms) [inline pkg/mesh-networking/apply/applier.go:693]: i believe this can panic if `mode` is nil; can we test?

File: api/networking/v1/virtual_mesh.proto (hunk 4 above)
Comment (harveyxia) [reply api/networking/v1/virtual_mesh.proto:112]: good call, i forgot that `oneofs` don't show up in generated docs

**Review by ilackarms — APPROVED** [review]
